File: python/arrays-and-hashing/sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        two, zero = len(nums) - 1, 0
        for i in range(len(nums)):
            if i > two:
                return

            while nums[i] == 2:
                nums[i], nums[two] = nums[two], nums[i]
                two -= 1

                if i > two:
                    return

            while nums[i] == 0:
                nums[i], nums[zero] = nums[zero], nums[i]
                zero += 1

                if zero > i:
                    break

File: c++/arrays-and-hashing/sort-colors.cc
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int n = nums.size();
        int left = 0;
        int right = n - 1;
        for (int i = 0; i <= right; ++i) {
            while (nums[i] == 2) {
                nums[i] = nums[right];
                nums[right] = 2;
                right--;

                if (i > right) {
                    return;
                } 
            }
            
            while (nums[i] == 0) {
                nums[i] = nums[left];
                nums[left] = 0;
                left++;

                if (left > i) {
                    break;
                }
            }
        }
    }
};
